flocker/node/agents/cinder.py

```diff
@@ -457,37 +487,140 @@ def allocation_unit(self):
         """
         return int(GiB(1).to_Byte().value)
 
-    def compute_instance_id(self):
+    def _metadata_from_configdrive(self):
+        """
+        Attempt to mount the configdrive and return the decoded metadata file.
+        """
+        mountpoint = FilePath(mkdtemp(suffix='.flocker.node.agents.cinder'))
+        metadata_file = mountpoint.descendant(
+            ['openstack', 'latest', 'meta_data.json']
+        )
+        try:
+            try:
+                result = check_output(
+                    ['blkid', '-l', '-t', 'LABEL=config-2', '-o', 'device'],
+                    stderr=STDOUT,
+                )
+            except CalledProcessError as e:
+                if e.returncode == 2 and not e.output:
+                    Message.new(
+                        message_type=(
+                            u"flocker:node:agents:blockdevice:openstack:"
+                            u"compute_instance_id:configdrive_not_available"),
+                        error_message=unicode(e),
+                    ).write()
+                    return
+                raise
+            device_path = FilePath(result.rstrip())
+            check_output(
+                ['mount', '-o', 'ro', device_path.path, mountpoint.path],
+                stderr=STDOUT,
+            )
+            try:
+                try:
+                    content = metadata_file.getContent()
+                except IOError as e:
+                    Message.new(
+                        message_type=(
+                            u"flocker:node:agents:blockdevice:openstack:"
+                            u"compute_instance_id:metadata_file_not_found"),
+                        error_message=unicode(e),
+                    ).write()
+                    return
+                return json.loads(content)
+            finally:
+                check_output(
+                    ['umount', mountpoint.path],
+                    stderr=STDOUT,
+                )
+        finally:
+            mountpoint.remove()
+
+    def _compute_instance_id_from_configdrive(self):
+        """
+        Attempt to mount the config drive and get instance ID from the metadata
+        file.
+
+        http://docs.openstack.org/user-guide/cli-config-drive.html
+        """
+        metadata = self._metadata_from_configdrive()
+        if metadata is not None:
+            return metadata["uuid"]
+
+    def _compute_instance_id_from_metadata_service(self):
+        """
+        Attempt to access instance ID from the metadata service.
+
+        http://docs.openstack.org/admin-guide/compute-networking-nova.html#metadata-service
+        """
+        try:
+            response = requests.get(
+                'http://169.254.169.254/openstack/latest/meta_data.json',
```

Comment:
i dont know much about the metadata service but is `latest` the version of openstack or version of metadata?